landoapi/api/uplift.py

```diff
@@ -23,49 +24,89 @@
-            "target_repository": data["repository"],
-        },
-    )
-    is_approval, revision, target_repository = check_approval_state(
-        g.phabricator,
-        revision_id=data["revision_id"],
-        target_repository_name=data["repository"],
-    )
-    logger.info(
-        "Approval state is valid",
-        extra={"style": "approval" if is_approval else "uplift"},
-    )
+    if not repository.approval_required:
+        return problem(
+            400,
+            f"Repository {repo_name} is not an uplift repository.",
+            "Please select an uplift repository to create the uplift request.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400",
+        )
 
     try:
-        if is_approval:
-            output = create_approval_request(
-                g.phabricator, revision, data["form_content"]
-            )
-        else:
-            output = create_uplift_revision(
-                g.phabricator, revision, target_repository, data["form_content"]
+        logger.info(
+            "Checking approval state",
+            extra={
+                "revision": revision_id,
+                "target_repository": repo_name,
+            },
+        )
+        revision_data, target_repository = get_uplift_conduit_state(
+            phab,
+            revision_id=revision_id,
+            target_repository_name=repo_name,
+        )
+        logger.info("Approval state is valid")
+    except ValueError as err:
+        return problem(
+            404,
+            "Revision not found",
+            err.args[0],
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404",
+        )
+
+    # Get the `#release-managers` group PHID for requesting review.
+    relman_phid = get_relman_group_phid(phab)
+    if not relman_phid:
+        return problem(
+            404,
+            "#release-managers group not found.",
+            (
+                "The RelMan review group could not be found. This is a server-side "
+                "issue, please file a bug."
+            ),
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404",
+        )
+
+    try:
+        output = {}
+
+        parent_phid = None
+        for revision in revision_data.revisions.values():
+            # Get the relevant diff.
+            diff_phid = phab.expect(revision, "fields", "diffPHID")
+            diff = revision_data.diffs[diff_phid]
+
+            # Create the revision.
+            rev = create_uplift_revision(
+                phab, revision, diff, parent_phid, relman_phid, target_repository
             )
+            new_uplift_phid = rev["revision_phid"]
+
+            # Store the returned revision information for the response.
+            output[new_uplift_phid] = rev
+
+            # Set `parent_phid` so the next created revision has this one set as the parent.
+            parent_phid = new_uplift_phid
+
     except Exception as e:
```

Comment:
We should log a little more info here, especially since it's possible that there are stray revisions or diffs or other data that would have been created but not to completion. We should also abandon or mark any revisions that were created in the process so that they don't cause pollution. Ideally we do some clean up as well but at least all revisions/diffs that were partially created should be logged in the error.

This could also be refactored so that the transactions and API calls are collected and then called separately (e.g. outside the for loop) and updated afterwards with any relevant new info as needed, making error handling a little more robust.

Reply:
Switched the implementation to maintain a stack of previously created revisions instead of just setting `parent_phid`. Then we get the previous parent with `stack[-1]["phid"]` and we can log the stack if we hit a failure creating a revision.